pyrdp/player/Mp4EventHandler.py

```diff
@@ -0,0 +1,151 @@
+        :param fps: The frame rate (30 recommended).
+        :param progress: An optional callback (sig: `() -> ()`) whenever a frame is muxed.
+        """
+
+        self.sink = Mp4Sink()
+        self.filename = filename
+        self.mp4 = f = av.open(filename, 'w')
+        self.stream = f.add_stream('h264', rate=fps)
+        self.stream.pix_fmt = 'yuv420p'
+        self.progress = progress
+        self.scale = False
+        self.mouse = (0, 0)
+        self.fps = fps
+        self.delta = 1000 // fps  # ms per frame
+        self.log = logging.getLogger(__name__)
+        self.log.info('Begin MP4 export to %s: %d FPS', filename)
+        self.timestamp = self.prevTimestamp = None
+
+        super().__init__(self.sink)
+
+    def onPDUReceived(self, pdu: PlayerPDU):
+        super().onPDUReceived(pdu)
+
+        # Make sure the rendering surface has been created.
+        if self.sink.screen is None:
+            return
+
+        ts = pdu.timestamp
+        self.timestamp = ts
+
+        if self.prevTimestamp is None:
+            dt = self.delta
+        else:
+            dt = self.timestamp - self.prevTimestamp  # ms
+        nframes = (dt // self.delta)
+        if nframes > 0:
+            for _ in range(nframes):
+                self._writeFrame(self.sink.screen)
+            self.prevTimestamp = ts
+            self.log.debug('Rendered %d still frame(s)', nframes)
+
+    def cleanup(self):
+        # Add one second worth of padding so that the video doesn't end too abruptly.
+        for _ in range(self.fps):
+            self._writeFrame(self.sink.screen)
+
+        self.log.info('Flushing to disk: %s', self.filename)
+        for pkt in self.stream.encode():
+            if self.progress:
+                self.progress()
+            self.mp4.mux(pkt)
+        self.log.info('Export completed.')
+        self.mp4.close()
+
+    def onMousePosition(self, x, y):
+        self.mouse = (x, y)
+        super().onMousePosition(x, y)
+
+    def onCapabilities(self, caps):
+        bmp = caps[CapabilityType.CAPSTYPE_BITMAP]
+        (w, h) = (bmp.desktopWidth, bmp.desktopHeight)
+        self.sink.resize(w, h)
+
+        if w % 2 != 0:
+            self.scale = True
+            w += 1
+        if h % 2 != 0:
+            self.scale = True
+            h += 1
+
+        self.stream.width = w
+        self.stream.height = h
+
+        super().onCapabilities(caps)
+
+    def onFinishRender(self):
+        # When the screen is updated, always write a frame.
+        self.prevTimestamp = self.timestamp
+        self._writeFrame(self.sink.screen)
+
+    def _writeFrame(self, surface: QImage):
+        w = self.stream.width
+        h = self.stream.height
+        surface = surface.scaled(w, h) if self.scale else surface
+        frame = av.VideoFrame.from_image(ImageQt.fromqimage(surface))
+
+        # Draw the mouse pointer.
+        # NOTE: We could render mouse clicks by changing the color of the brush.
```

Comment:
I didn't put it as a TODO because I hate leaving rotting TODOs in the code. I just wanted to leave a note for anyone reading the code that this would be a good place for this improvement. I also feel like this is too minor to open an idea issue about (feels like scope creep TBH).

If you don't agree with NOTE I'll just remove the entire comment.

Reply:
I don't feel that this is scope creep but it's not important either. Leave it as NOTE, this is not worth arguing over :slightly_smiling_face: